### PR TITLE
Cloud spanner templates, tests and doco

### PR DESCRIPTION
**Pull request by sourced-vince**

### CODE REVIEW

File: community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        delete_config
+    fi
+}
+
+@test "Creating deployment ${DEPLOYMENT_NAME} from ${CONFIG}" {
+    run gcloud deployment-manager deployments create "${DEPLOYMENT_NAME}" \
+        --config ${CONFIG} \
+        --project "${CLOUD_FOUNDATION_PROJECT_ID}"
+    [[ "$status" -eq 0 ]]
+}
```
Comment (fawix) [inline community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: I'm getting the following error when running the tests:

```
 ✗ Creating deployment ocsi-courseralab1-guestbook-cloud-spanner-77y078y02m from .ocsi-courseralab1-guestbook-cloud-spanner-77y078y02m.yaml
   (in test file templates/cloud_spanner/tests/integration/cloud_spanner.bats, line 55)
     `[[ "$status" -eq 0 ]]' failed
```

I don't see where `status` is being set; did you mean `$?` ?

File: community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    [[ "$status" -eq 0 ]]
+    [[ "$output" =~ "test-myspannercluster-${RAND}" ]]
+}
+
+@test "Verifying that Spanner cluster IAM was created as part of ${DEPLOYMENT_NAME}" {
+    run gcloud spanner instances get-iam-policy test-myspannercluster-"${RAND}" \
+        --project "${CLOUD_FOUNDATION_PROJECT_ID}"
+    [[ "$status" -eq 0 ]]
+    [[ "$output" =~ "${PROJECT_NUMBER}@cloudservices.gserviceaccount.com" ]]
+}
```
Comment (fawix) [inline community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: This one is also failing, 

```
 ✗ Verifying that Spanner cluster IAM was created as part of ocsi-courseralab1-guestbook-cloud-spanner-77y078y02m
   (in test file templates/cloud_spanner/tests/integration/cloud_spanner.bats, line 69)
     `[[ "$output" =~ "${PROJECT_NUMBER}@cloudservices.gserviceaccount.com" ]]' failed
```

DM logs are:

```shell
ocsi-courseralab1-guestbook-cloud-spanner-77y078y02m has resource warnings
test-myspannercluster-77y078y02m-setIamPolicy: {"ResourceType":"gcp-types/spanner-v1:spanner.projects.instances.setIamPolicy","ResourceErrorCode":"400","ResourceErrorMessage":{"code":400,"message":"Invalid service account email (@cloudservices.gserviceaccount.com).","status":"INVALID_ARGUMENT","statusMessage":"Bad Request","requestPath":"https://spanner.googleapis.com/v1/projects/ocsi-courseralab1-guestbook/instances/test-myspannercluster-77y078y02m:setIamPolicy","httpMethod":"POST"}}
```

File: community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    [[ "$status" -eq 0 ]]
+    [[ "$output" =~ "spannerdb1" ]]
+}
+
+@test "Verifying that Spanner DB IAM was created as part of ${DEPLOYMENT_NAME}" {
+    run gcloud spanner databases get-iam-policy spannerdb1 --instance test-myspannercluster-"${RAND}" \
+        --project "${CLOUD_FOUNDATION_PROJECT_ID}"
+    [[ "$status" -eq 0 ]]
+    [[ "$output" =~ "role: roles/spanner.databaseAdmin" ]]
+}
```
Comment (fawix) [inline community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: And this one too:

```
 ✗ Verifying that Spanner DB IAM was created as part of ocsi-courseralab1-guestbook-cloud-spanner-77y078y02m
   (in test file templates/cloud_spanner/tests/integration/cloud_spanner.bats, line 83)
     `[[ "$output" =~ "role: roles/spanner.databaseAdmin" ]]' failed
```

File: community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats (hunk 1 above)
Comment (sourced-vince) [reply community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: I verified the test and they work for me.

Looks like the deployment failed and according to the DM logs, specifically the setting spanner IAM. Can you check if you have `roles/spanner.admin` permissions?

Comment (fawix) [reply community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: Ok, I investigated some more and it looks like the `${PROJECT_NUMBER}` is empty resulting in the following YAML file:

```yaml
 imports:
   - path: templates/cloud_spanner/cloud_spanner.py
     name: cloud_spanner.py
   
   resources:
     - name: test-myspannercluster-5simlkgtbz
       type: cloud_spanner.py
       properties:
         displayName: "Spanner Cluster 1"
         nodeCount: 2
         instanceConfig: nam3
         bindings:
           - role: "roles/spanner.admin"
             members:
               - "serviceAccount:@cloudservices.gserviceaccount.com"
         databases:
           - name: "spannerdb1"
             bindings:
               - role: "roles/spanner.databaseAdmin"
                 members:
                   - "serviceAccount:@cloudservices.gserviceaccount.com"
```

It makes sense given the DM error message:

```shell
..."ResourceErrorCode":"400","ResourceErrorMessage":{"code":400,"message":"Invalid service account email (@cloudservices.gserviceaccount.com).","status":"INVALID_ARGUMENT","statusMessage":"Bad
    Request",...
```

Is there any config I need to add to run this template?

Comment (sourced-vince) [reply community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: Found the problem. Please try with the latest commit.

Comment (fawix) [reply community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: Hi Vince, not yet - still failing. This is what I get now:

```shell
The fingerprint of the deployment is SepRNUJAbtV-1y96ebgZrA==
Waiting for create [operation-1540936892736-57979541ce200-199f3a06-b818546e]...
.......................failed.
ERROR: (gcloud.deployment-manager.deployments.create) Error in Operation [operation-1540936892736-57979541ce200-199f3a06-b818546e]: errors:
- code: RESOURCE_ERROR
  location: /deployments/ocsi-courseralab1-guestbook-cloud-spanner-16g41lij4a/resources/test-myspannercluster-16g41lij4a-setIamPolicy
  message: '{"ResourceType":"gcp-types/spanner-v1:spanner.projects.instances.setIamPolicy","ResourceErrorCode":"400","ResourceErrorMessage":{"code":400,"message":"Service
    account creator@cloudservices.gserviceaccount.com does not exist.","status":"INVALID_ARGUMENT","statusMessage":"Bad
    Request","requestPath":"https://spanner.googleapis.com/v1/projects/ocsi-courseralab1-guestbook/instances/test-myspannercluster-16g41lij4a:setIamPolicy","httpMethod":"POST"}}'
```

Generated YAML looks like this:

```yaml
imports:
   - path: templates/cloud_spanner/cloud_spanner.py
     name: cloud_spanner.py
   
   resources:
     - name: test-myspannercluster-16g41lij4a
       type: cloud_spanner.py
       properties:
         displayName: "Spanner Cluster 1"
         nodeCount: 2
         instanceConfig: nam3
         bindings:
           - role: "roles/spanner.admin"
             members:
               - "serviceAccount:creator@cloudservices.gserviceaccount.com"
         databases:
           - name: "spannerdb1"
             bindings:
               - role: "roles/spanner.databaseAdmin"
                 members:
                   - "serviceAccount:creator@cloudservices.gserviceaccount.com"
```

Comment (fawix) [reply community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats]: PS: to be more specific these are the test results:
```shell
 ✓ Verifying that Spanner cluster was created as part of ocsi-courseralab1-guestbook-cloud-spanner-16g41lij4a
 ✗ Verifying that Spanner cluster IAM was created as part of ocsi-courseralab1-guestbook-cloud-spanner-16g41lij4a
   (in test file templates/cloud_spanner/tests/integration/cloud_spanner.bats, line 70)
     `[[ "$output" =~ "${PROJECT_NUMBER}@cloudservices.gserviceaccount.com" ]]' failed
 ✓ Verifying that Spanner DB was created as part of ocsi-courseralab1-guestbook-cloud-spanner-16g41lij4a
 ✗ Verifying that Spanner DB IAM was created as part of ocsi-courseralab1-guestbook-cloud-spanner-16g41lij4a
   (in test file templates/cloud_spanner/tests/integration/cloud_spanner.bats, line 84)
     `[[ "$output" =~ "role: roles/spanner.databaseAdmin" ]]' failed
 ✓ Deleting deployment
```

**Comment by fawix** [comment]
Still doesn't work but I found another issue - when exporting project number:
```bash
export PROJECT_NUMBER=$(gcloud projects list | grep "${CLOUD_FOUNDATION_PROJECT_ID}" | awk '{ print $3}')
```

My project name has a space so awk doesn't work here:
```shell
↪ gcloud projects list | grep $CLOUD_FOUNDATION_PROJECT_ID                             
ocsi-courseralab1-guestbook              ocsi-project creator               461314075586

↪ gcloud projects list | grep $CLOUD_FOUNDATION_PROJECT_ID | awk '{ print $3}'            
creator
```

It's better to get the last string in the phrase  like so:

```shell
↪ gcloud projects list | grep $CLOUD_FOUNDATION_PROJECT_ID | awk {'print $NF'}          
461314075586
```
I've already changed and tested.

**Pull request merged**